pkg/clusteragent/autoscaling/cluster/controller.go

```diff
@@ -0,0 +1,430 @@
+
+	autoscalingWorkqueue := workqueue.NewTypedRateLimitingQueueWithConfig(
+		workqueue.DefaultTypedItemBasedRateLimiter[string](),
+		workqueue.TypedRateLimitingQueueConfig[string]{
+			Name:            subsystem,
+			MetricsProvider: autoscalingQueueMetricsProvider,
+		},
+	)
+
+	baseController, err := autoscaling.NewController(controllerID, c, dynamicClient, dynamicInformer, nodePoolGVR, isLeaderFunc, store, autoscalingWorkqueue)
+	if err != nil {
+		return nil, err
+	}
+
+	c.Controller = baseController
+
+	// TODO add later, if needed, when adding more telemetry
+	// store.RegisterObserver(autoscaling.Observer{
+	// 	DeleteFunc: unsetTelemetry,
+	// })
+
+	c.store = store
+
+	return c, nil
+}
+
+// PreStart is called before the controller starts
+func (c *Controller) PreStart(ctx context.Context) {
+	startLocalTelemetry(ctx, c.localSender, []string{"kube_cluster_id:" + c.clusterID})
+}
+
+// Process implements the Processor interface (so required to be public)
+// this processes what's in the workqueue, comes from the store or cluster
+func (c *Controller) Process(ctx context.Context, key, ns, name string) autoscaling.ProcessResult {
+	// Follower should not process workqueue items
+	if !c.IsLeader() {
+		return autoscaling.ProcessResult{}
+	}
+
+	// Try to get NodePool from cluster
+	np := &karpenterv1.NodePool{}
+	npUnstr, err := c.Lister.Get(name)
+	if err == nil {
+		err = autoscaling.FromUnstructured(npUnstr, np)
+	}
+
+	switch {
+	case apierrors.IsNotFound(err):
+		// Ignore not found error as it will be created later
+		np = nil
+	case err != nil:
+		log.Errorf("Unable to retrieve NodePool: %w", err)
+		return autoscaling.Requeue
+	case npUnstr == nil:
+		log.Errorf("Could not parse empty NodePool from local cache")
+		return autoscaling.Requeue
+	}
+
+	// Track if found NodePool is fully managed
+	fullyManaged := false
+	if np != nil {
+		fullyManaged = isCreatedByDatadog(np.GetLabels())
+	}
+
+	// TODO create duplicate NodePools with greater weight, rather than updating user NodePools
+	mnp, foundInStore := c.store.LockRead(name, true)
+
+	if foundInStore {
+		if np == nil {
+			// Present in store but not found in cluster; create it
+			if err = c.createNodePool(ctx, mnp); err != nil {
+				log.Errorf("Error creating NodePool: %v", err)
+			}
+		} else {
+			// Present in store and found in cluster; update it
+			// TODO check if hash of spec from remote config matches current object before updating
+			if err = c.patchNodePool(ctx, np, mnp); err != nil {
+				log.Errorf("Error updating NodePool: %v", err)
+			}
+		}
+	} else {
+		if fullyManaged {
+			// Not present in store, and the cluster NodePool is fully managed, then delete the NodePool
+			if err = c.deleteNodePool(ctx, name); err != nil {
+				log.Errorf("Error deleting NodePool: %v", err)
+			}
+		} else {
+			// Not present in store and the cluster NodePool is not fully managed, do nothing
+			log.Debugf("NodePool not found in store and is not fully managed: %s", name)
```

Comment:
nit: in the log can we explicitly say we are doing nothing to make it clearer that no action is being taken? like:
```suggestion
			log.Debugf("NodePool: %s not found in store and is not fully managed, nothing to do", name)
```